### PR TITLE
GitHub Issue NOAA-EMC/GSI#308. Add observation provider info into obs diagnostic file for ps/w/spd/q/t

### LINKED ISSUE #308: Updated observer code(setup*.f90) for ps/w/spd/q/t to add observation provider info into obs diag files
Adding modifications to obs setup code(setup*.f90) for ps/q/t/w/spd, to include the observation provider and sub-provider information into the obsdiags filee. Adding new option "l_obsprvdiag" to trigger the output. These provider information can be read with read_diag_conv and used in the AutoObsQC for 3DRTMA.

Also modifying the code in read_diag_conv.f90 under util/Analysis_Utilities/read_diag to read the provider information in the obs-diag file and write out in text format which can be used for auto-obsqc.

### PR DESCRIPTION
**Pull request by hu5970**
GitHub issue NOAA-EMC/GSI#308: Add observation provider info into obs diagnostic file for ps/w/spd/q/t

    * Adding modifications to obs setup code(setup*.f90) for ps/q/t/w/spd, to include the observation provider and
    sub-provider information into the diag files, which will be used in read_diag_conv_3drtma in Auto-ObsQC

    * Also adding a new specific code read_diag_conv_3drtma.f90 under read_diag to read the obs-diag file which includes
    the obs provider information for auto-qc.

### CODE REVIEW

**Comment by MichaelLueken-NOAA** [comment]
@hu5970 - Please let me know who you would like to have review your changes.  Thanks!

**Comment by MichaelLueken-NOAA** [comment]
Tagging @ManuelPondeca-NOAA to review these changes.

**Comment by hu5970** [comment]
@GangZhao-NOAA @MatthewMorris-NOAA Gang and Matt, please review this PR, thanks, Ming

**Review by MatthewMorris-NOAA — APPROVED** [review]

**Review by GangZhao-NOAA — APPROVED** [review]
Everything looks good to me.
I give approval to this PR.
Thanks!

 -Gang

**Review by MichaelLueken — COMMENTED** [review]
Hi Ming,

There are four minor issues that need to be addressed before I can submit this work to the review committee:

Please add proper precision to -9999. in setupq.f90 and setupt.f90

File: src/gsi/setupq.f90 [hunk 1]
```diff
@@ -1201,8 +1218,43 @@ subroutine contents_binary_diagp_
...
+           enddo
+           do jj=1,miter
+              ioff=ioff+1
+              rdiagbufp(ioff,iip) = odiag%obssen(jj)
+           enddo
+        endif
+
+        if (twodvar_regional .or. l_obsprvdiag) then
+           ioff = ioff + 1
+           rdiagbufp(ioff,iip) = -9999. !  data(idomsfc,i) ! dominate surface type
```
[the remark this answers is not in the document]
Comment (hu5970) [reply src/gsi/setupq.f90:1247]: Added

File: src/gsi/setupt.f90 [hunk 2]
```diff
@@ -1604,8 +1620,44 @@ subroutine contents_binary_diagp_
...
+        enddo
+        do jj=1,miter
+           idia=idia+1
+           rdiagbufp(idia,iip) = odiag%obssen(jj)
+        enddo
+     endif
+
+    if (twodvar_regional .or. l_obsprvdiag) then
+       idia = idia + 1
+       rdiagbufp(idia,iip) = -9999. ! data(idomsfc,i) ! dominate surface type
```
[the remark this answers is not in the document]
Comment (hu5970) [reply src/gsi/setupt.f90:1649]: Added

File: src/gsi/setupt.f90 [hunk 3]
```diff
@@ -1604,8 +1620,44 @@ subroutine contents_binary_diagp_
...
+       idia = idia + 1
+       rdiagbufp(idia,iip) = -9999. ! data(izz,i)     ! model terrain at observation location
```
[the remark this answers is not in the document]
Comment (hu5970) [reply src/gsi/setupt.f90:1651]: added

**Comment by hu5970** [comment]
@MichaelLueken-NOAA Thanks for reviewing. I have updated code based on your comments. Ming

**Comment by MichaelLueken-NOAA** [comment]
@hu5970 I see that you forced pushed changes to the save_obsProvider branch, but I'm still seeing the real values without precision.  I have cloned your fork and checked out the save_obsProvider branch, and see that the real values are still missing precision.

When you made the change, did you git add (or git commit -a --amend) before using git commit --amend?  I'm trying to see why a successful push didn't update setupq.f90 and setupt.f90 in your branch.

**Comment by hu5970** [comment]
@MichaelLueken-NOAA I must do something wrong when I did the push. I have did it again and this time it works. Please check. Thanks, Ming

**Comment by MichaelLueken-NOAA** [comment]
@hu5970 I checked and everything looks good.  I should now be able to submit your work to the review committee.  Thanks!  Mike

**Comment by MichaelLueken-NOAA** [comment]
The due date has passed with no feedback from the review committee, so I will now give final approval to this update and merge it to the authoritative repository.

**Pull request merged**